plugins/inputs/tail/tail.go

```diff
@@ -86,89 +91,88 @@ func (t *Tail) Start(acc telegraf.Accumulator) error {
 	defer t.Unlock()
 
 	t.acc = acc
-	t.tailers = make(map[string]*tail.Tail)
-
-	return t.tailNewFiles(t.FromBeginning)
-}
-
-func (t *Tail) tailNewFiles(fromBeginning bool) error {
-	var seek *tail.SeekInfo
-	if !t.Pipe && !fromBeginning {
-		seek = &tail.SeekInfo{
-			Whence: 2,
-			Offset: 0,
-		}
-	}
+	t.lines = make(chan *logline.LogLine)
+	t.files = make(map[string]bool)
 
 	var poll bool
 	if t.WatchMethod == "poll" {
 		poll = true
 	}
 
-	// Create a "tailer" for each file
+	var err error
+	t.watcher, err = watcher.NewLogWatcher(t.PollInterval.Duration, !poll, watcher.Logger(taillog.DiscardingLogger))
+	if err != nil {
+		defer close(t.lines)
+		return err
+	}
+
+	opts := []tailer.Option{tailer.Logger(taillog.DiscardingLogger)}
+	if t.FromBeginning {
+		opts = append(opts, tailer.OneShot)
+	}
+
+	t.tailer, err = tailer.New(t.lines, t.watcher, opts...)
+	if err != nil {
+		defer t.watcher.Close()
+		defer close(t.lines)
+		return err
+	}
+
+	parser, err := t.parserFunc()
+	if err != nil {
+		defer t.tailer.Close()
+		return err
+	}
+	t.wg.Add(1)
+	go t.parse(parser)
+
+	return t.tailNewFiles(t.FromBeginning)
+}
+
+func (t *Tail) tailNewFiles(fromBeginning bool) error {
+
 	for _, filepath := range t.Files {
 		g, err := globpath.Compile(filepath)
 		if err != nil {
 			t.acc.AddError(fmt.Errorf("E! Error Glob %s failed to compile, %s", filepath, err))
 		}
 		for _, file := range g.Match() {
-			if _, ok := t.tailers[file]; ok {
+			if _, ok := t.files[file]; ok {
 				// we're already tailing this file
 				continue
 			}
 
-			tailer, err := tail.TailFile(file,
-				tail.Config{
-					ReOpen:    true,
-					Follow:    true,
-					Location:  seek,
-					MustExist: true,
-					Poll:      poll,
-					Pipe:      t.Pipe,
-					Logger:    tail.DiscardingLogger,
-				})
+			err := t.tailer.TailPath(file)
```

Comment:
If the file doesn't exist this returns an error, normally globpath.Match only returns matches, but if no globs are in use it just returns the path.  Even if we changed Match then its possible that the file would go away before we can tail, so I think we need to check if the file is not found or permission error occurs here and just ignore it.

I think ideally this function could return one of these known errors https://golang.org/pkg/os/#pkg-variables

Reply:
The tail lib handles glob patterns itself; perhaps I remove the globbing code from the plugin and rely on the tail library.
https://github.com/sgtsquiggs/tail/blob/master/tailer/tail.go#L152-L181

I wrote it the way it is now so that I could easily track which files are being watched in the plugin. I believe a better way would be to have the tail lib expose callbacks for when a file is watched/unwatched. Additionally it would allow the plugin to accurately report when a file is deleted.

Reply:
W/r to the errors; the tail lib is [wrapping](https://godoc.org/github.com/pkg/errors#Wrapf) the os errors with a custom error; I can check for underlying error and ignore.